lean4-infoview/src/infoview/rpcSessions.tsx

```diff
@@ -0,0 +1,195 @@
+    uri: DocumentUri
+}
+
+interface RpcConnected {
+    uri: DocumentUri
+    sessionId: string
+}
+
+interface RpcCallParams extends TextDocumentPositionParams {
+    sessionId: string
+    method: string
+    params: any
+}
+
+interface RpcReleaseParams {
+    uri: DocumentUri
+    sessionId: string
+    ref: RpcPtr<any>
+}
+
+const RpcNeedsReconnect = -32900
+
+class RpcSession {
+    #ec: EditorConnection
+    #uri: DocumentUri
+    #finalizers: FinalizationRegistry<RpcPtr<any>>
+
+    constructor(readonly sessionId: string, uri: DocumentUri, ec: EditorConnection) {
+        this.#ec = ec
+        this.#uri = uri
+        // Here we hook into the JS GC and send release-reference requests whenever
+        // the GC finalizes an `RpcPtr`. Requires ES2021.
+        this.#finalizers = new FinalizationRegistry(ptr => {
+            const params: RpcReleaseParams = {
+                uri: this.#uri,
+                sessionId: this.sessionId,
+                ref: ptr,
+            }
+
+            // TODO(WN): batch multiple of these into a single notif
+            void this.#ec.api.sendClientNotification('$/lean/rpc/release', params)
+        })
+    }
+
+    async call(pos: DocumentPosition, method: string, params: any): Promise<any> {
+        const rpcParams: RpcCallParams = {
+            ...DocumentPosition.toTdpp(pos),
+            sessionId: this.sessionId,
+            method: method,
+            params: params,
+        }
+        return this.#ec.api.sendClientRequest('$/lean/rpc/call', rpcParams)
+            .then(val => {
+                // const s = JSON.stringify(val)
+                // console.log(`'${method}(${JSON.stringify(params)})' at '${pos.line}:${pos.character}' -> '${s.length < 200 ? s : '(..)'}'`)
+                return val
+            })
+    }
+
+    registerRef(ptr: RpcPtr<any>) {
+        this.#finalizers.register(ptr, RpcPtr.copy(ptr))
+    }
+}
+
+/** Provides an interface for making RPC calls to the Lean server. */
+export class RpcSessions {
+    /** Maps each URI to the connected RPC session, if any, at the corresponding file worker. */
+    #connected: Map<DocumentUri, RpcSession>
+    /** Like `#connected`, but for sessions which are currently connecting. */
+    #connecting: Map<DocumentUri, Promise<RpcSession>>
+    #ec: EditorConnection
+    setSelf: (_: (_: RpcSessions) => RpcSessions) => void
+
+    constructor(ec: EditorConnection) {
+        this.#connected = new Map()
+        this.#connecting = new Map()
+        this.#ec = ec
+        this.setSelf = () => { }
+    }
+
+    private async sessionAt(uri: DocumentUri): Promise<RpcSession | undefined> {
+        if (this.#connected.has(uri)) return this.#connected.get(uri)!
+        else if (this.#connecting.has(uri)) return this.#connecting.get(uri)!
+        else return undefined
+    }
+
+    private connectAt(uri: DocumentUri): void {
+        // NOTE(WN): a request would be slightly simpler to handle than a notification,
```

Comment:
I think a request is the better choice here, if explicit initialization is necessary.  (And doesn't seem to be any harder to implement in the server, whether you call `handleRpcConnect` in `handleRequest` or `handleNotification` shouldn't make a big difference.)

Reply:
The issue is that in the server we currently maintain that notifications are handled synchronously and can modify server state while requests run in thread-pools and cannot. We'd need a special case for this one request, or a bigger overhaul to allow synchronous requests in general. It's doable but I'd rather keep it as-is for now given that the notification works fine.